### PR TITLE
Introduce options.check_flush_compaction_key_order

### PR DESCRIPTION
**Pull request by siying**
Summary:
Introduce an new option options.check_flush_compaction_key_order, by default set to true, which checks key order of flush and compaction, and fail the operation if the order is violated.
Also did minor refactor hash checking code, which consolidates the hashing logic to a vlidation class, where the key ordering logic is added.

Test Plan: Add unit tests to validate the check can catch reordering in flush and compaction, and can be properly disabled.

### CODE REVIEW

File: db/output_validator.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  icmp_(_icmp), enable_order_check_(_enable_order_check), enable_hash_(_enable_hash) {}
+  Status Add(const Slice& key, const Slice& value) {
+    if (enable_hash_) {
+      // Generate a rolling 64-bit hash of the key and values
+      paranoid_hash_ =
+          Hash64(key.data(), key.size(), paranoid_hash_);
+      paranoid_hash_ =
+          Hash64(value.data(), value.size(), paranoid_hash_);
+    }
+    if (enable_order_check_) {
```
Comment (riversand963) [inline db/output_validator.h:29]: I feel that if we enable hash check, then we should also enable order check since its overhead is relatively low.

File: db/compaction/compaction_job.cc [hunk 2]
```diff
@@ -1593,14 +1607,17 @@ Status CompactionJob::OpenCompactionOutputFile(
...
-    out.paranoid_hash = 0;
-    sub_compact->outputs.push_back(out);
+    FileMetaData meta;
+    meta.fd = FileDescriptor(file_number,
+                             sub_compact->compaction->output_path_id(), 0);
+    meta.oldest_ancester_time = oldest_ancester_time;
+    meta.file_creation_time = current_time;
+    sub_compact->outputs.emplace_back(
+        std::move(meta), cfd->internal_comparator(),
+        /*enable_order_check*/
```
Comment (riversand963) [inline db/compaction/compaction_job.cc:1617]: `/*enable_order_check=*/`

File: db/builder.cc [hunk 3]
```diff
@@ -96,7 +97,11 @@ Status BuildTable(
          column_family_name.empty());
   // Reports the IOStats for flush for every following bytes.
   const size_t kReportFlushIOStatsEvery = 1048576;
-  uint64_t paranoid_hash = 0;
+  OutputValidator output_validator(
+      internal_comparator,
+      /*enable_order_check*/
```
Comment (riversand963) [inline db/builder.cc:102]: `/*enable_order_check=*/`

File: db/corruption_test.cc [hunk 4]
```diff
@@ -619,6 +622,78 @@ TEST_F(CorruptionTest, ParanoidFileChecksOnCompact) {
...
+  ASSERT_NOK(s);
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->DisableProcessing();
+}
+
+TEST_F(CorruptionTest, DisableKeyOrderCheck) {
+  Options options;
+  ASSERT_OK(db_->SetOptions({{"check_flush_compaction_key_order", "false"}}));
+  DBImpl* dbi = static_cast_with_check<DBImpl>(db_);
+
+  SyncPoint::GetInstance()->SetCallBack(
```
Comment (riversand963) [inline db/corruption_test.cc:683]: Nit: maybe clear all callbacks after the test.

File: db/output_validator.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+// and the class validates key order and optionally calculate a hash
+// of all the key and value.
+class OutputValidator {
+ public:
+  explicit OutputValidator(const InternalKeyComparator& _icmp,
+                           bool _enable_order_check, bool _enable_hash) :
+  icmp_(_icmp), enable_order_check_(_enable_order_check), enable_hash_(_enable_hash) {}
+  Status Add(const Slice& key, const Slice& value) {
+    if (enable_hash_) {
+      // Generate a rolling 64-bit hash of the key and values
```
Comment (riversand963) [inline db/output_validator.h:23]: Just curious, any plan to make the hash algorithm configurable?

**Review by riversand963 — APPROVED** [review]
Thanks @siying for the PR. LGTM with a few minor comments.

**Comment by riversand963** [comment]
hmm, the test failures seem legit and need be fixed.

File: db/output_validator.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+// Pass every key/value of the file using OutputValidator::Add()
+// and the class validates key order and optionally calculate a hash
+// of all the key and value.
+class OutputValidator {
+ public:
+  explicit OutputValidator(const InternalKeyComparator& _icmp,
+                           bool _enable_order_check, bool _enable_hash) :
+  icmp_(_icmp), enable_order_check_(_enable_order_check), enable_hash_(_enable_hash) {}
+  Status Add(const Slice& key, const Slice& value) {
+    if (enable_hash_) {
```
Comment (mrambacher) [inline db/output_validator.h:22]: Can we move the implementation into a C++ file instead of a header and clean up the includes?

File: db/memtable.cc [hunk 7]
```diff
@@ -375,6 +375,7 @@ class MemTableIterator : public InternalIterator {
     PERF_COUNTER_ADD(next_on_memtable_count, 1);
     assert(Valid());
     iter_->Next();
+    TEST_SYNC_POINT_CALLBACK("MemTableIterator::Next:0", iter_);
```
Comment (mrambacher) [inline db/memtable.cc:378]: Just curious as to what this has to do with this change?

File: db/corruption_test.cc [hunk 8]
```diff
@@ -619,6 +622,78 @@ TEST_F(CorruptionTest, ParanoidFileChecksOnCompact) {
...
 
+TEST_F(CorruptionTest, CompactionKeyOrderCheck) {
+  Options options;
+  options.paranoid_file_checks = false;
+  options.create_if_missing = true;
+  options.check_flush_compaction_key_order = false;
+  Status s;
+  delete db_;
+  db_ = nullptr;
+  s = DestroyDB(dbname_, options);
```
Comment (mrambacher) [inline db/corruption_test.cc:633]: Should this status be asserted?

File: db/compaction/compaction_job.cc [hunk 9]
```diff
@@ -1275,8 +1289,8 @@ Status CompactionJob::FinishCompactionOutputFile(
       auto kv = tombstone.Serialize();
       assert(lower_bound == nullptr ||
              ucmp->Compare(*lower_bound, kv.second) < 0);
-      sub_compact->AddToBuilder(kv.first.Encode(), kv.second,
-                                paranoid_file_checks_);
+      sub_compact->AddToBuilder(kv.first.Encode(), kv.second)
+          .PermitUncheckedError();
```
Comment (mrambacher) [inline db/compaction/compaction_job.cc:1293]: Don't you need to check this Status to see out of order keys?

File: db/memtable.cc (hunk 7 above)
Comment (siying) [reply db/memtable.cc:378]: I didn't find a good way to fake a memtable that generates out of order keys. If there is a better way, will do it.

File: db/output_validator.h (hunk 5 above)
Comment (siying) [reply db/output_validator.h:23]: The hash is not exposed to users, unless the hash is not good enough I don't see a reason to make it configerable.

File: db/output_validator.h (hunk 1 above)
Comment (siying) [reply db/output_validator.h:29]: The order check is likely to be more expensive than hash check. It needs to make a copy to every key, which is not cheap.

File: db/output_validator.h [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+      // Generate a rolling 64-bit hash of the key and values
+      paranoid_hash_ =
+          Hash64(key.data(), key.size(), paranoid_hash_);
+      paranoid_hash_ =
+          Hash64(value.data(), value.size(), paranoid_hash_);
+    }
+    if (enable_order_check_) {
+      TEST_SYNC_POINT_CALLBACK("OutputValidator::Add:order_check", /*arg=*/nullptr);
+      if (key.size() < 8) {
```
[the remark this answers is not in the document]
Comment (siying) [reply db/output_validator.h:31]: Will do.

File: db/compaction/compaction_job.cc [hunk 11]
```diff
@@ -1276,8 +1290,8 @@ Status CompactionJob::FinishCompactionOutputFile(
       auto kv = tombstone.Serialize();
       assert(lower_bound == nullptr ||
              ucmp->Compare(*lower_bound, kv.second) < 0);
-      sub_compact->AddToBuilder(kv.first.Encode(), kv.second,
-                                paranoid_file_checks_);
+      // Range tombstone is not supported by output validator yet.
+      sub_compact->builder->Add(kv.first.Encode(), kv.second);
```
Comment (ajkr) [inline db/compaction/compaction_job.cc:1294]: This fixes the problem with `paranoid_file_checks` + `DeleteRange()`, right? @mrambacher would it solve the problem to extract this into a separate PR that we can easily backport to 6.12 and 6.13?

**Review by ajkr — APPROVED** [review]
Awesome!
